pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,950 @@
+			}
+			if !reflect.DeepEqual(oldSyncStatus, newSyncStatus) {
+				newSyncStatus.LastTransitionTime = metav1.Now()
+			}
+			newSyncStatuses = append(newSyncStatuses, newSyncStatus)
+		}
+	}
+
+	return
+}
+
+func getOldSyncStatus(syncSet CommonSyncSet, syncSetStatuses []hiveintv1alpha1.SyncStatus) (hiveintv1alpha1.SyncStatus, int) {
+	for i, status := range syncSetStatuses {
+		if status.Name == syncSet.AsMetaObject().GetName() {
+			return syncSetStatuses[i], i
+		}
+	}
+	return hiveintv1alpha1.SyncStatus{}, -1
+}
+
+func (r *ReconcileClusterSync) applySyncSet(
+	syncSet CommonSyncSet,
+	resourceHelper resource.Helper,
+	logger log.FieldLogger,
+) (
+	resourcesApplied []hiveintv1alpha1.SyncResourceReference,
+	resourcesInSyncSet []hiveintv1alpha1.SyncResourceReference,
+	requeue bool,
+	returnErr error,
+) {
+	resources, referencesToResources, decodeErr := decodeResources(syncSet, logger)
+	referencesToSecrets := referencesToSecrets(syncSet)
+	resourcesInSyncSet = append(referencesToResources, referencesToSecrets...)
+	if decodeErr != nil {
+		returnErr = decodeErr
+		return
+	}
+
+	applyFn := resourceHelper.Apply
+	applyFnMetricsLabel := labelApply
+	switch syncSet.GetSpec().ApplyBehavior {
+	case hivev1.CreateOrUpdateSyncSetApplyBehavior:
+		applyFn = resourceHelper.CreateOrUpdate
+		applyFnMetricsLabel = labelCreateOrUpdate
+	case hivev1.CreateOnlySyncSetApplyBehavior:
+		applyFn = resourceHelper.Create
+		applyFnMetricsLabel = labelCreateOnly
+	}
+
+	// Apply Resources
+	for i, resource := range resources {
+		returnErr, requeue = r.applyResource(i, resource, referencesToResources[i], applyFn, applyFnMetricsLabel, logger)
+		if returnErr != nil {
+			resourcesApplied = referencesToResources[:i]
+			return
+		}
+	}
+	resourcesApplied = referencesToResources
+
+	// Apply Secrets
+	for i, secretMapping := range syncSet.GetSpec().Secrets {
+		returnErr, requeue = r.applySecret(syncSet, i, secretMapping, referencesToSecrets[i], applyFn, applyFnMetricsLabel, logger)
+		if returnErr != nil {
+			resourcesApplied = append(resourcesApplied, referencesToSecrets[:i]...)
+			return
+		}
+	}
+	resourcesApplied = append(resourcesApplied, referencesToSecrets...)
+
+	// Apply Patches
+	for i, patch := range syncSet.GetSpec().Patches {
+		returnErr, requeue = r.applyPatch(i, patch, resourceHelper, logger)
+		if returnErr != nil {
+			return
+		}
+	}
+
+	logger.Info("syncset applied")
+	return
+}
+
+func decodeResources(syncSet CommonSyncSet, logger log.FieldLogger) (
+	resources []*unstructured.Unstructured, references []hiveintv1alpha1.SyncResourceReference, returnErr error,
+) {
+	var decodeErrors []error
+	for i, resource := range syncSet.GetSpec().Resources {
+		u := &unstructured.Unstructured{}
+		if err := yaml.Unmarshal(resource.Raw, u); err != nil {
+			logger.WithField("resourceIndex", i).WithError(err).Warn("error decoding unstructured object")
+			decodeErrors = append(decodeErrors, errors.Wrapf(err, "failed to decode resource %d", i))
```

Comment:
If you get a decode error, should you continue processing the syncset? 
For example, I have a syncset with mode=sync that contains a certain number of resources. I update the syncset and make a mistake while editing one of the resources, makes it impossible to decode. As it stands, it looks like the condition would be reported, but also my resource which was already applied to the target cluster would be deleted because it's no longer considered part of the resources in the syncset.

Maybe it should be a validation error instead?

Reply:
If you make a mistake during editing, then you cannot rely on Hive to catch that for you. Let's say that the mistake you made was mistyping the name of your resource. The resource with the name that you were using is going to be deleted. In practice, the only decode error that is going to happen is a missing Kind field. If it is malformed json, then there will be an error at update time.

Reply:
can't we just do the unstructured decode at validation though? It would then only leave existing data with potential decode errors.

Reply:
OK. I will do the decode in validation.